WIP: CEG-Prophecy implementation in Pono

makaimann opened the pull request.

makaimann commented:
Hey @ahmed-irfan, just FYI I have the first working version running (finally). I was able to prove a very simple property (the motivating example) over bit-vectors. 

However, one thing that I realized is that k-induction will almost never work. I was thinking that it would be able to prove this property, but it actually can't because of the form of the prophecized property. The invariant that we need is actually over the prophecy variable, even though that predicate never appears directly in the system. I think we can address this for some simple cases, (if the relevant index is a free variable in the property). But, I'm not sure how important that is -- Interpolant-based model checking is able to pick up this example by finding a nice interpolant over UF.

I'll start tests soon and keep you updated.

makaimann commented:
Looks like Travis fails on two of the configurations. I’ll look into this more tomorrow. It apparently is triggered by using a different compiler.

ahmed-irfan commented on engines/ceg_prophecy_arrays.cpp:
```diff
@@ -0,0 +1,459 @@
+        return ProverResult::FALSE;
+      }
+      reached_k_++;
+    } while (num_added_axioms_);
```
when will it exit from the loop? (only in the case of !refine())

makaimann replied:
Either for `!refine()` or when no axioms were added (e.g. `num_added_axioms_ == 0`). So it tries to refine bmc until no new axioms were found, or there's a concrete counterexample.

ahmed-irfan commented on engines/ceg_prophecy_arrays.cpp:
```diff
@@ -0,0 +1,459 @@
+ProverResult CegProphecyArrays::prove()
+{
+  ProverResult res = ProverResult::FALSE;
+  while (res == ProverResult::FALSE) {
```
why are we refining before checking the abstract property?

makaimann replied:
Ahh it's just a heuristic -- we used that in the prototype as well. It adds a few axioms first before trying to prove it. If the property doesn't require any axioms, then it will refine quickly anyway and get to the underlying model checker.

ahmed-irfan commented on engines/ceg_prophecy_arrays.cpp:
```diff
@@ -0,0 +1,459 @@
+}
+
+ProverResult CegProphecyArrays::prove()
+{
```
also, maybe use check_until with INT_MAX

makaimann replied:
Makes sense! That means it will call `check_until(INT_MAX)` for the underlying model checker instead of `prove`. So far that means exactly the same thing, but I wasn't sure it would stay that way. I'll make the change, but just thought I'd point that out.

ahmed-irfan commented on engines/ceg_prophecy_arrays.cpp:
```diff
@@ -0,0 +1,459 @@
+    return false;
+  }
+
+  UnorderedTermSet consecutive_axioms = aae_.get_consecutive_axioms();
```
```suggestion
  const UnorderedTermSet &consecutive_axioms = aae_.get_consecutive_axioms();
```

makaimann replied:
Good idea. This is actually a little tricky, because `consecutive_axioms` might get updated. If we found nonconsecutive axioms, then we search for consecutive axioms again after refining with auxiliary variables. However, the reference will point to fresh data so it should be okay. I added comments about that.

makaimann replied:
Oh actually, it's also convenient to have a separate copy so that `reduce_consecutive_axioms` can modify it in-place. I could make this change and also change the interface to `reduce_consecutive_axioms` if you prefer?

ahmed-irfan replied:
let's do it later, as you suggested.

ahmed-irfan commented on engines/ceg_prophecy_arrays.cpp:
```diff
@@ -0,0 +1,459 @@
+  }
+
+  UnorderedTermSet consecutive_axioms = aae_.get_consecutive_axioms();
+  AxiomVec nonconsecutive_axioms = aae_.get_nonconsecutive_axioms();
```
```suggestion
  const AxiomVec &nonconsecutive_axioms = aae_.get_nonconsecutive_axioms();
```

makaimann replied:
This is the same as the one above. It's copied so that it can be modified in place when being reduced. I could change the interface of the reducers, although then we'd have to have two versions -- the reduced one and the unreduced one. As it is now, it just gets edited when reduced. Thoughts?

ahmed-irfan reviewed (commented):
Gave a pass. 
Please see the comments

makaimann commented on modifiers/abstractor.h:
```diff
@@ -0,0 +1,80 @@
+  // getters
+  const TransitionSystem & conc_ts() const { return conc_ts_; };
+
+  TransitionSystem & abs_ts() const { return abs_ts_; };
```
Ahh yeah so the abs_ts was actually non-const on purpose, because we need to add auxiliary variables and constraints to it later. In the future, we could make it const but just copy it if you think that's better. But for now, that requires this PR which hasn't been merged yet: https://github.com/upscale-project/pono/pull/88. I think we should definitely discuss the abstraction refinement interface more.

makaimann commented on modifiers/array_abstractor.cpp:
```diff
@@ -0,0 +1,418 @@
+  abs_rts.set_init(abs_init);
+  abs_rts.set_trans(abs_trans);
+
+  // TODO: remove these debug prints
```
Let's just do it now, thanks!

makaimann commented on modifiers/prophecy_modifier.cpp:
```diff
@@ -0,0 +1,42 @@
+  string name = "proph_" + target->to_string() + "_" + std::to_string(delay);
+  Term proph_var = ts_.make_statevar(name, target->get_sort());
+  // make it frozen
+  ts_.assign_next(proph_var, proph_var);
```
Yep, it still supports it!

makaimann commented on pono.cpp:
```diff
@@ -86,7 +106,9 @@ int main(int argc, char ** argv)
   logger.set_verbosity(pono_options.verbosity_);
 
-  try {
+  // TEMP comment this out for development
```
Yep, thanks. It does make sense to catch any exceptions, right? Or do you think it's better to let them go uncaught?

makaimann commented:
Thanks for the review @ahmed-irfan! I know it was a lot. I addressed some your comments in the latest commit and commented about the rest. Let me know your thoughts when you get a chance.

makaimann commented on refiners/array_axiom_enumerator.cpp:
```diff
@@ -0,0 +1,842 @@
+  }
+
+  // populate axioms
+  // TODO: use an unsat core to prune
```
Ope this is already being done now, I'll remove the TODO (see labels for unsat core membership below)

makaimann commented on refiners/array_axiom_enumerator.cpp:
```diff
@@ -0,0 +1,842 @@
+  collect_arrays_and_indices();
+  create_lambda_indices();
+
+  // TODO: remove these debug prints
```
done

makaimann commented on refiners/array_axiom_enumerator.cpp:
```diff
@@ -0,0 +1,842 @@
+  af.visit(trans);
+  af.visit(conc_bad_);
+
+  // TODO: create lambda variables!!
```
Nope, just an old TODO -- removed it. It gets called elsewhere. Thanks for catching that!

makaimann commented on refiners/array_axiom_enumerator.cpp:
```diff
@@ -0,0 +1,842 @@
+  // must be within bound
+  assert(i <= bound_);
+
+  UnorderedTermSet & indices = only_curr ? cur_index_set_ : index_set_;
```
It could contain inputs.

makaimann commented on refiners/array_axiom_enumerator.cpp:
```diff
@@ -0,0 +1,842 @@
+  size_t num_found_lemmas;
+  Term unrolled_ax;
+  for (AxiomInstantiation ax_inst : index_axioms(ac, unrolled_indices)) {
+    if (only_curr && !ts_.only_curr(ax_inst.ax)) {
```
Same, it could contain inputs.

makaimann commented on refiners/array_axiom_enumerator.cpp:
```diff
@@ -0,0 +1,842 @@
+  //       version of axioms correctly
+  //       not explicitly calling next here -- is that a problem?
+  //       should be okay as long as we add next version of axioms
+  //       that are only over state variables
```
Removed the TODO

makaimann commented on refiners/array_axiom_enumerator.cpp:
```diff
@@ -0,0 +1,842 @@
+    // in the axiom
+    size_t max_k = ts_.only_curr(ax_inst.ax) ? bound_ : bound_ - 1;
+    for (size_t k = 0; k <= max_k; ++k) {
+      unrolled_ax = un_.at_time(ax_inst.ax, k);
```
Ahh right, that's a little tricky. So the indices have already been unrolled, so that's where the other time comes in.

ahmed-irfan replied:
i see. Maybe add a simple example in the comment

makaimann replied:
Good idea! Added a simple example with 931a15940312f3fec36e2bde91287cb073a9a9b6

ahmed-irfan reviewed (approved):
Thanks for the great work.

Overall LGTM.
One minor request to add an example of non-consecutive axiom instantiation. (see the comment)

ahmed-irfan commented:
@makaimann need to resolve conflicts
